Dev 1

sergsnmail opened the pull request:
Добрый день.
Домашнее задание 3

Описание проделанной работы в файле [TaskReport.txt](https://github.com/sergsnmail/jl4fcloud/blob/dev-1/TaskReport.txt) в корне проекта

sergsnmail committed:
add architecture.txt

sergsnmail committed:
First dev commit (add TaskReport.txt)

Commit:
Переработаны (упрощена) модель объектов сообщений. В ней удалены лишние классы MethodRequest.java, MethodRequestParameter.java, MethodResponse.java, MethodResponseResult.java.
Добавлены методы сообщений PutFilesMethod и GetFilesMethod.

sergsnmail committed:
1. Решена проблема при передачи длинных тектстовых сообщений (в начало pipleLine добавлени обработчик  new LineBasedFrameDecoder(5000 * 1024,true,false)).
2. Добавлен пакет filemanager на клиенте. Реализует функционал фрагментации и передачи файлов от клиента на сервер (в тестовом виде).
3. Добавлен класс Base64Converter для конвертации byte[] -> string и string -> byte[]
4. В интерфейсе добавлена индикация передачи файла (в тестовом виде).

Lezenford commented on netty-client/src/main/java/client/network/ClientNetwork.java:
```diff
@@ -0,0 +1,78 @@
+                                new LineBasedFrameDecoder(5000 * 1024,true,false),
+                                new StringDecoder(),
+                                new StringEncoder(),
+                                new JsonObjectDecoder(),
```
этот обработчик лишний если у вас отработает LineBasedFrameDecoder

sergsnmail replied:
Я его держу, чтобы он валидировал входную строку JSON, поскольку я установил довольно избыточный размер фрейма.  И вообще он странный: по идее он не должен пропускать не валидное сообщение дальше по пайпу, но я столкнулся с ситуацией, когда мое сообщение JSON приходило двумя кусками, и этот декодер пропустил дальше невалидный JSON (количество открытых скобок не равно количеству закрытых), соответственно, дальше выпадало исключение джексона. Именно после этого, я поставил размер фрейма и ограничил размер передаваемого куска файла таким, что полученный JSON гарантированно помещался в размер фрейма.

Lezenford commented on netty-common-obj/src/main/java/message/Request.java:
```diff
@@ -0,0 +1,84 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Request extends Message {
```
у объекта всего 2 поля, билдер не делает ничего более валидации полей на null. Для такого небольшого объекта билдер обычно излишнее усложнение

sergsnmail replied:
Первоначально билдер был по сложнее. После упрощения классов модели сообщений, он подсократился, но я не стал его убирать совсем. В последних коммитах, если он по прежнему не измениться, я от него избавлюсь.

Lezenford commented on netty-common-obj/src/main/java/message/Response.java:
```diff
@@ -0,0 +1,68 @@
+import message.common.Method;
+import message.common.Result;
+
+public class Response extends Message {
```
Аналогично. 1 поле и билдер к нему при наличии конструктора

Lezenford commented on netty-server/src/main/java/db/DbStorage.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    public Connection getConnection() throws ClassNotFoundException, SQLException {
+        Class.forName("org.sqlite.JDBC");
```
что делает эта строка? ее результат нигде не используется

sergsnmail replied:
во всех примерах по работе sqlite эта строка инициализирует драйвер JDBC. Загружает статическую инициализацию.

Lezenford commented on netty-server/src/main/java/server/handler/MessageServerHandler.java:
```diff
@@ -0,0 +1,169 @@
+            GetFilesMethod method = (GetFilesMethod) request.getMethod();
+            GetFilesParam param = method.getParameter();;
+            GetFilesResult result = new GetFilesResult();
+            UserSession session = request.getSession();
```
не совсем понимаю зачем вы передаете сессию в части сообщения - каждое открытое соединение имеет свой набор классов в пайплайне, т.е. у каждого клиента своя копия хэндлера. После того как прошла авторизация вы можете просто положиться на механизм нетти и записать в сам хэндлер юзера за которым он закреплен. тогда не потребуется каждый раз тянуть информацию из сообщения, т.к. у каждой открытой сессии свой экземпляр и всю информацию можно хранить там

sergsnmail replied:
Небольшой закос по стейтлесс сессию. Но, думаю, я это переделаю так как вы посоветовали.

Lezenford reviewed (commented):
Отличная работа! практически полнофункциональное приложение. Отличный способ работы с json в плане полиморфизма.

sergsnmail committed:
1. Переработан механизм пакетной передачи файлов.

sergsnmail committed:
2. Переработан механизм пакетной передачи файлов. Передача вынесена в отдельный класс. Информация о файле вынесена в отдельный класс.

sergsnmail committed:
3. Реализован механизм отслеживания файлов в указанной папке. Реализован механизм хранения файлов на сервере и регистрация их в БД. Настроен мавен проект для клиента FX. Переработана структура пакетов.

sergsnmail committed:
4. Добавлен отчет о проделанной работе (файл TaskReport.txt).

sergsnmail committed:
5. Переработан механизм передачи файлов между клиентом и сервером. Теперь он не сьедает память при передачи большого количества файлов.
6. Доработан механизм слежения за папкой синхронизации. Теперь при копировании в папку синхронизации большого количества файлов они все попадают в очередь для пердачи.
7. Добавлен новый метод для получения информации о файле на сервере (GetFileInfo)

Commit:
6. Упрощен пайп-лайн на клиенте и сервере. JSON конвертируется сразу в байты, минуя строку. В связи с этим удалены кодеки String. Также изменены MessageEncoder и MessageDecoder.
7. В пайп-лайн в начало включены LengthFieldBasedFrameDecoder и LengthFieldPrepender.
8. В код проекта вставлены поясняющие комментарии, а также удалены мусорные строки.
9. Исправленые небольшие неточности по работе клиента с сервером.
10. Полностью переработан механизм передачи файлов на сервер от клиента. Предыдущий вариант удален.
11. Немного доработано окно клиента для корректной обработки ресайза окна.

Текущие итоги:
Клиент
 - Работа клиента осуществляется только по направлению к серверу. Т.е клиент корректно отправляет файлы из каталога синхронизации на сервер.
 - Корректкно реагирует на добавление каталогов в папке синхронизации.
 - Корректно отрабатывает при массовом копировании файлов в каталог синхронизации.
 - Корректно передает на сервер файлы большого размера (проверялось на файлах 7-8ГБ).
 - Корректно осуществляет индикацию передачи файлов в интерфейсе пользователя (прогресс-бар, количество отправленных пакетов/ общее количество пакетов, имя отправляемого файла).
 - Отправляет данные для регистрации нового пользователя.
Сервер
 - Регистрирует нового пользователя
 - Предоставляет информацию по запрошенному файлу
 - Отправляет список файлов пользователя клиенту
 - Оптимально организует хранение файлов в каталоге операционной системы.
 - Принимает файлы от клиента и регистрирует в БД

Commit:
12. Добавлен отчет о проделанной работе (файл TaskReport.txt).

Lezenford commented on netty-client/src/main/java/com/sergsnmail/client/transfer/PackageCollection.java:
```diff
@@ -0,0 +1,125 @@
+            /**
+             * ожидание разблокировки файла
+             */
+            while(!isLocked){
```
вы храните состояние в локальной переменной. никто кроме этого треда в один момент все равно не сможет занять этот lock, к чему он в итоге?

Lezenford commented on netty-common-obj/src/main/java/com/sergsnmail/common/json/JSONConverter.java:
```diff
@@ -0,0 +1,47 @@
+        {
+            e.printStackTrace();
+        }
+        return null;
```
подобная обработка легко спровоцирует дальше nullpointer, не уверен что это хорошая идея. Возможно стоит заменить исключение на runtime, чтобы не оборачивать вызов этого мапера каждый раз в tryCache, но глобально (на уровне хэндлера, например) оно будет перехватываться и закрывать канал

sergsnmail committed:
13. Добавлены параметры запуска сервера.
13.1 Изменен цикл ожидания разблокировки файла в PackageCollection
13.2 Доработка логики отправки файла на сервер

Commit:
14. Переработан механизм отслеживания файлов в папке синхронизации. Теперь при массовом копировании в папку синхронизации определяется событие копирования, и проводится дополнительное сканирование папки в которой было обнаружено копирование. Данное поведение необходимо из-за того, что при копировании FileWatcher не успевает зарегистрировать слушатель на папку, и некоторые файлы успевают скопироваться до регистрации прослушивания.
14.1 При указании папки синхронизации начинается автоматическое копирование всех файлов с сервера. !!!Данное поведение необходимо доработать с возможностью проверки существования файлов на диске!!!
14.2 Переработан интерфейс клиента, теперь отображается, как процесс загрузки, так и отправки фалов на сервер.

Lezenford commented on netty-client/src/main/java/com/sergsnmail/client/network/transfer/TransferWorker.java:
```diff
@@ -0,0 +1,189 @@
+    @Override
+    public void run() {
+        try {
+            while (!Thread.currentThread().isInterrupted()) {
```
не обязательно проверять на interrupt. Если вы далее используете sleep, то проверка на интерапт будет при каждом вызове этого метода. Тогда можно писать лаконичное while(true) и спрятать выход из цикла в cache

sergsnmail committed:
15. Дополнены парметры запуска сервера. Теперь можно указать путь до файла БД на диске, параметр -db.
15.1 Добавлены параметры запуска клиента (-host, -port).
15.2 Доработан механизм предварительного сканирования файлов в папке синхронизации. Теперь при указании папки будет производится полное однократое ее сканирование на наличие текущего содержимого.

sergsnmail committed:
16 Доработан механизм предварительного сканирования файлов в папке синхронизации. Теперь при отсутствии активности в папке синхронизации поток WatchRepo засыпает.
16. Настроена сборка проекта. Полная сборка осуществляется по команде mvn clean package
